// react-hot-keys/cjs/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _hotkeysJs = _interopRequireDefault(require("hotkeys-js"));

var ReactHotkeys = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ReactHotkeys, _React$Component);

  var _super = (0, _createSuper2.default)(ReactHotkeys);

  function ReactHotkeys(props) {
    var _this;

    (0, _classCallCheck2.default)(this, ReactHotkeys);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isKeyDown", false);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handle", void 0);
    _this.onKeyDown = _this.onKeyDown.bind((0, _assertThisInitialized2.default)(_this));
    _this.onKeyUp = _this.onKeyUp.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleKeyUpEvent = _this.handleKeyUpEvent.bind((0, _assertThisInitialized2.default)(_this));
    _this.handle = {};
    return _this;
  }

  (0, _createClass2.default)(ReactHotkeys, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props = this.props,
          filter = _this$props.filter,
          splitKey = _this$props.splitKey;

      if (filter) {
        _hotkeysJs.default.filter = filter;
      }

      _hotkeysJs.default.unbind(this.props.keyName);

      (0, _hotkeysJs.default)(this.props.keyName, {
        splitKey: splitKey
      }, this.onKeyDown);
      document && document.body.addEventListener('keyup', this.handleKeyUpEvent);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      _hotkeysJs.default.unbind(this.props.keyName);

      this.isKeyDown = true;
      this.handle = {};
      document && document.body.removeEventListener('keyup', this.handleKeyUpEvent);
    }
  }, {
    key: "onKeyUp",
    value: function onKeyUp(e, handle) {
      var _this$props2 = this.props,
          onKeyUp = _this$props2.onKeyUp,
          disabled = _this$props2.disabled;
      !disabled && onKeyUp && onKeyUp(handle.shortcut, e, handle);
    }
  }, {
    key: "onKeyDown",
    value: function onKeyDown(e, handle) {
      var _this$props3 = this.props,
          onKeyDown = _this$props3.onKeyDown,
          allowRepeat = _this$props3.allowRepeat,
          disabled = _this$props3.disabled;
      if (this.isKeyDown && !allowRepeat) return;
      this.isKeyDown = true;
      this.handle = handle;
      !disabled && onKeyDown && onKeyDown(handle.shortcut, e, handle);
    }
  }, {
    key: "handleKeyUpEvent",
    value: function handleKeyUpEvent(e) {
      if (!this.isKeyDown) return;
      this.isKeyDown = false;
      if (this.props.keyName && this.props.keyName.indexOf(this.handle.shortcut) < 0) return;
      this.onKeyUp(e, this.handle);
      this.handle = {};
    }
  }, {
    key: "render",
    value: function render() {
      return this.props.children || null;
    }
  }]);
  return ReactHotkeys;
}(_react.default.Component);

exports.default = ReactHotkeys;
(0, _defineProperty2.default)(ReactHotkeys, "defaultProps", {
  filter: function filter(event) {
    var target = event.target || event.srcElement;
    var tagName = target.tagName;
    return !(target.isContentEditable || tagName === 'INPUT' || tagName === 'SELECT' || tagName === 'TEXTAREA');
  }
});
(0, _defineProperty2.default)(ReactHotkeys, "propTypes", {
  keyName: PropTypes.string,
  filter: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyUp: PropTypes.func,
  disabled: PropTypes.bool,
  splitKey: PropTypes.string
});
module.exports = exports.default;
